### PR TITLE
[VL] Fix the compile problem with dependencies installed in user directories

### PR DESCRIPTION
**Pull request by jackylee-ch**
## What changes were proposed in this pull request?
In our scenario, we use a non-root user to compile this project, and all dependencies are installed in user directory, including gcc12.1. Thus we need to support passing the dependencies into cmake, otherwise no dependency can be found in system dir.

### CODE REVIEW

**Comment by zhejiangxiaomai** [comment]
Hi, @jackylee-ch  thanks for your contribution. But I see some errors in CI job (https://github.com/oap-project/gluten/actions/runs/4239989250/jobs/7368562949), would you fix it. I mean if user do not specify the some library. like FMT_LIBRARIES.

**Comment by jackylee-ch** [comment]
@zhejiangxiaomai Thanks for your attension. I'm working on the CI Errors, I will resolve it.

**Comment by jinchengchenghh** [comment]
Our internal jenkins performance test use non-root user, but it works well, not sure it is platform issue

**Comment by jinchengchenghh** [comment]
I know the background, usually users use apt install to install dependency to /usr, but in your environment, you have the specified lib directory

**Comment by zhejiangxiaomai** [comment]
Would you add a feature that if we cannot find needed libs(like glog、re2、gflags、libunwind, snappy、zstd), we can build then from source.   here is a example used in meta/velox. (https://github.com/facebookincubator/velox/blob/main/CMakeLists.txt#L45-L51)

**Comment by jackylee-ch** [comment]
> Would you add a feature that if we cannot find needed libs(like glog、re2、gflags、libunwind, snappy、zstd), we can build then from source. here is a example used in meta/velox. (https://github.com/facebookincubator/velox/blob/main/CMakeLists.txt#L45-L51)

Yeah, the best way to resolve this problem is to add the [resolve_dependency](https://github.com/facebookincubator/velox/blob/main/ThirdpartyToolchain.cmake#L369) method. It will try to find package first, then build from source.

**Comment by zhejiangxiaomai** [comment]
> > Would you add a feature that if we cannot find needed libs(like glog、re2、gflags、libunwind, snappy、zstd), we can build then from source. here is a example used in meta/velox. (https://github.com/facebookincubator/velox/blob/main/CMakeLists.txt#L45-L51)
> 
> Yeah, the best way to resolve this problem is to add the [resolve_dependency](https://github.com/facebookincubator/velox/blob/main/ThirdpartyToolchain.cmake#L369) method. It will try to find package first, then build from source. If we agree with this way, I can remove the extra `include_directories` and `target_link_libraries`, and only left `Findfolly` is this pr. Then I can move on the solution of `resolve_dependency` in other prs.

Thanks.

**Review by zhejiangxiaomai — APPROVED** [review]
look good to me.

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -84,7 +84,6 @@ add_compile_options(-Wall)
 add_compile_options(-Wno-sign-compare)
 add_compile_options(-Wno-comment)
 
-add_compile_options(-Werror)
```
Comment (zuochunwei) [inline cpp/CMakeLists.txt:87]: why remove this option?

Comment (jackylee-ch) [reply cpp/CMakeLists.txt:87]: There was some error throwed while compiling with gcc12.1

File: cpp/core/CMakeLists.txt [hunk 2]
```diff
@@ -111,6 +111,7 @@ macro(build_protobuf)
     PROPERTIES IMPORTED_LOCATION "${PROTOBUF_STATIC_LIB}" INTERFACE_INCLUDE_DIRECTORIES
                "${PROTOBUF_INCLUDE_DIR}")
   add_dependencies(protobuf::libprotobuf protobuf_ep)
+  include_directories(${PROTOBUF_INCLUDE_DIR})
```
Comment (zhztheplayer) [inline cpp/core/CMakeLists.txt:114]: Can we use `target_include_directories`?
https://cmake.org/cmake/help/latest/command/target_include_directories.html

File: cpp/velox/CMakeLists.txt [hunk 3]
```diff
@@ -189,8 +189,18 @@ set_target_properties(velox PROPERTIES
     LIBRARY_OUTPUT_DIRECTORY ${root_directory}/releases
     )
 
+## If the folly is not builded in system directories, please add
+## `-DCMAKE_PREFIX_PATH="${folly builded directory}" to the parameters of cmake.
+## It is also applicable to other dependencies.
 find_package(Folly REQUIRED CONFIG)
 find_package(gflags REQUIRED COMPONENTS shared CONFIG)
+find_package(GTest REQUIRED)
+find_package(Protobuf REQUIRED)
```
Comment (zhztheplayer) [inline cpp/velox/CMakeLists.txt:198]: Do we have to find protobuf again? Wasn't it found in cpp/gluten/CMakeLists.txt?

File: cpp/core/CMakeLists.txt (hunk 2 above)
Comment (jackylee-ch) [reply cpp/core/CMakeLists.txt:114]: Sure, I will try it.

File: cpp/velox/CMakeLists.txt (hunk 3 above)
Comment (jackylee-ch) [reply cpp/velox/CMakeLists.txt:198]: Hm, I didn't find it in cpp/gluten/CMakeLists.txt. After my tests, we still need to find it here and include `Protobuf_INCLUDE_DIRS`.

Comment (zhztheplayer) [reply cpp/velox/CMakeLists.txt:198]: Sorry it's in `cpp/core`

https://github.com/oap-project/gluten/blob/a04106623966f8cb384a2c64d89c02fbe18f0eb1/cpp/core/CMakeLists.txt#L116-L124

We had a flag `BUILD_PROTOBUF`. I am not sure whether `find_package(protobuf)` still works when `BUILD_PROTOBUF=ON`?

Comment (zhztheplayer) [reply cpp/velox/CMakeLists.txt:198]: > After my tests, we still need to find it here and include Protobuf_INCLUDE_DIRS.

And I assume if target `gluten` requires protobuf headers publicly in CMake, then the headers can be propagated to target `velox`. You can have a try on it.

Comment (jackylee-ch) [reply cpp/velox/CMakeLists.txt]: > We had a flag BUILD_PROTOBUF. I am not sure whether find_package(protobuf) still works when BUILD_PROTOBUF=ON?

Maybe we can use `CACHE PATH` to solve this problem. User can `BUILD_PROTOBUF` to choose whether they want build protobuf, and the variable for protobuf is only configured in core/CMakeLists.txt only once.

**Comment by jackylee-ch** [comment]
@zhztheplayer @jinchengchenghh any more question about this pr?

**Review by zhztheplayer — APPROVED** [review]

**Pull request merged**